### DISCUSSION TITLE
Use DAC converter of  ESP32

### DISCUSSION THREAD

**Discussion by roesl** [arendst/Tasmota#17035] [Q&A] [answered]
Hello,

I need to simulate a temperatur sensor for heating control, because I dont be able to use direct connection to the sensor.
Is it possible to use the DAC feature of ESP32 to send a 8 bit value to the module and set the output to a defined analog value ?

**Comment by sfromis** [comment, marked as answer]
Yes, you can use the DAC feature [from Berry](https://tasmota.github.io/docs/Berry/#dac-gpios).

**Reply by sfromis** [reply]
Her's a rudimentary example of how you can implement a command to use the DAC support in Berry:
```
tasmota.add_cmd('DACVoltage',
  def (cmd, idx, millivolt)
    if idx == 1 || idx==2
      var pin = 27 - idx
      gpio.pin_mode(pin, gpio.DAC)
      gpio.dac_voltage(pin, int(millivolt))
      tasmota.resp_cmnd_done()
    else
      tasmota.resp_cmnd_error()
    end
  end)
```

**Reply by roesl** [reply]
Thank you, but how I have to configure the pin and how can I send the value with MQTT ? 
When I start the script in the console nothing matters

**Reply by sfromis** [reply]
The code example works without configuring, and it creates a command you can publish via MQTT, just like any other command. 

Unclear what you mean by "When I start the script in the console nothing matters". Don't expect anything in the Berry console, it simply registers a command. If you want to have it available after boot, you can put it in `autoexec.be`

**Reply by roesl** [reply]
can you please write the MQTT command which set the value ?
Is the configuration stored and available after reboot ?

**Reply by barbudor** [reply]
Obvioulsy @sfromis code creates a command named "DACVoltage" which accept 2 possible index values 1 or 2
So like for any tasmota command  you need something to publish the value as the payload to the topic `cmnd/topicofyourdevice/DACVoltage1` (for GPIO26) or `cmnd/topicofyourdevice/DACVoltage2` (for GPIO25) (replace `topicofyourdevice` by the appropriate value for your device)
The value should be in milliVolt as the code expressly states it

**Reply by barbudor** [reply]
If you execute the code in the Berry console it will not be ertain after reboot
You need to place the code in a file named `autoexec.be`

**Reply by roesl** [reply]
now it works partially - but the script reports an error

**Reply by sfromis** [reply]
If you want a suggestion, you need to tell us more than "an error". What exactly do you do, and what error do you get?

Did you get that the simplest would be for you to put the code sample into the file `autoexec.be` on the local Tasmota file system?

**Reply by roesl** [reply]
there is the file autoexec.bec with 

![autoexec bec](https://user-images.githubusercontent.com/23452986/201208213-fa2869af-28c4-4a33-9b1a-0469631fcc62.jpg)

**Reply by roesl** [reply]
and there is only DacVoltage2 at pin 25 working

**Reply by sfromis** [reply]
You neither told what error you got, nor what you did. Content of `autoexec.bec` is not at all interesting, it is just a "compiled" version of `autoexec.be`.

I did a small test, and saw both DACVoltage1 and DACVoltage2 working. If 1 and 2 are swapped for you, you can change the calculation of pin number to match what you expect. What **exactly** do you mean by "only DacVoltage2 at pin 25 working". What did you do, what did you expect to get, and what did you get instead?

**Reply by roesl** [reply]
Now all is fine - I measured the wrong pin :(
Thank you for your patience
